configuration/configurator/ConfiguratorAPI.cpp

```diff
@@ -0,0 +1,520 @@
+}
+
+void setIndex(const char *pXPath, int newIndex)
+{
+    assert(newIndex >= 0);
+
+    CRestriction *pRestriction = CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getRestrictionFromXPath(pXPath);
+
+    assert(pRestriction != NULL);
+    assert(pRestriction->getEnumerationArray() != NULL);
+    pRestriction->getEnumerationArray()->setEnvValueNodeIndex(newIndex);
+
+    CConfigSchemaHelper::getInstance()->setEnvTreeProp(pXPath, pRestriction->getEnumerationArray()->item(newIndex).getValue());
+}
+
+const char* getTableValue(const char *pXPath, int nRow)
+{
+    assert(pXPath != NULL && *pXPath != 0);
+
+    CAttribute *pAttribute = NULL;
+    CElement *pElement = NULL;
+
+    if (CConfigSchemaHelper::isXPathTailAttribute(pXPath) == true)
+       pAttribute = CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getAttributeFromXPath(pXPath);
+
+    if (pAttribute == NULL)
+    {
+        pElement =  CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getElementFromXPath(pXPath);
+        assert(pElement != NULL);
+
+        return pAttribute->getInstanceValue();
+    }
+    else
+    {
+        assert(pAttribute != NULL);
+        if (nRow == 1)
+            return pAttribute->getInstanceValue();
+        else
+        {
+            StringBuffer strXPath(pXPath);
+            const StringBuffer strXPathOriginal(pXPath);
+
+            int offset = strXPathOriginal.length() - (CConfigSchemaHelper::stripXPathIndex(strXPath) + 1) ;
+            CConfigSchemaHelper::stripXPathIndex(strXPath);
+
+            strXPath.appendf("[%d]", nRow);
+            strXPath.append(strXPathOriginal, offset, strXPathOriginal.length() - offset);
+            pAttribute =  CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getAttributeFromXPath(strXPath.str());
+
+            if (STRICTNESS_LEVEL >= DEFAULT_STRICTNESS)
+                assert(pAttribute != NULL);
+
+            if (pAttribute == NULL)
+                return NULL;
+
+            return pAttribute->getInstanceValue();
+        }
+    }
+}
+
+void setTableValue(const char *pXPath, int index, const char *pValue)
+{
+    UNIMPLEMENTED;
+}
+
+int getNumberOfUniqueColumns()
+{
+    return CConfigSchemaHelper::getInstance()->getEnvironmentXPathSize();
+}
+
+const char* getColumnName(int idx)
+{
+    if (idx < CConfigSchemaHelper::getInstance()->getEnvironmentXPathSize())
+        return CConfigSchemaHelper::getInstance()->getEnvironmentXPaths(idx);
+    else
+        return NULL;
+}
+
+int getNumberOfRows(const char* pXPath)
+{
+    assert(pXPath != NULL && *pXPath != 0);
+    PROGLOG("Get number of rows for %s = %d", pXPath, CConfigSchemaHelper::getInstance()->getElementArraySize(pXPath));
+
+    return CConfigSchemaHelper::getInstance()->getElementArraySize(pXPath);
+}
+
+int getNumberOfTables()
+{
+    return CConfigSchemaHelper::getInstance()->getNumberOfTables();
+}
+
+const char* getServiceName(int idx, char *pName)
```

Comment:
(if there's a reason you CAN'T use a StringBuffer, then you need to pass in the target buffer size and check that the length you are copying does not exceed it).